ml/randomforest.h

```diff
@@ -0,0 +1,657 @@
+        float weight;               ///< Weight of the example 
+        int exampleNum;             ///< index into feature array
+    };
+    
+    // Entry for an individual feature
+    struct Feature {
+        Feature()
+            : active(false), ordinal(true),
+              info(nullptr)
+        {
+        }
+
+        bool active;  ///< If true, the feature can be split on
+        bool ordinal; ///< If true, it's continuous valued; otherwise categ.
+        const DatasetFeatureSpace::ColumnInfo * info;
+        BucketList buckets;  ///< List of bucket numbers, per example
+    };
+
+    // All rows of data in this partition
+    std::vector<Row> rows;
+
+    // All features that are active
+    std::vector<Feature> features;
+
+    /** Reserve enough space for the given number of rows. */
+    void reserve(size_t n)
+    {
+        rows.reserve(n);
+    }
+
+    /** Add the given row. */
+    void addRow(const Row & row)
+    {
+        rows.push_back(row);
+    }
+
+    void addRow(bool label, float weight, int exampleNum)
+    {
+        rows.emplace_back(Row{label, weight, exampleNum});
+    }
+
+    template<typename Float>
+    struct WT {
+        WT()
+            : v { 0, 0 }
+        {
+        }
+
+        Float v[2];
+
+        Float & operator [] (bool i)
+        {
+            return v[i];
+        }
+
+        const Float & operator [] (bool i) const
+        {
+            return v[i];
+        }
+
+        bool empty() const { return total() == 0; }
+
+        Float total() const { return v[0] + v[1]; }
+
+        WT & operator += (const WT & other)
+        {
+            v[0] += other.v[0];
+            v[1] += other.v[1];
+            return *this;
+        }
+
+        WT operator + (const WT & other) const
+        {
+            WT result = *this;
+            result += other;
+            return result;
+        }
+
+        WT & operator -= (const WT & other)
+        {
+            v[0] -= other.v[0];
+            v[1] -= other.v[1];
+            return *this;
+        }
+
+        typedef Float FloatType;
+    };
+
+    //typedef WT<double> W;
+    typedef WT<ML::FixedPointAccum64> W;
+
+    /** Split the partition here. */
+    std::pair<PartitionData, PartitionData>
+    split(int featureToSplitOn, int splitValue, const W & wLeft, const W & wRight, const W & wAll)
+    {
+     //   std::cerr << "spliting on feature " << featureToSplitOn << " bucket " << splitValue << std::endl;
+
+        ExcAssertGreaterEqual(featureToSplitOn, 0);
+        ExcAssertLess(featureToSplitOn, features.size());
+
+        PartitionData sides[2];
+        PartitionData & left = sides[0];
+        PartitionData & right = sides[1];
+
+        left.fs = fs;
+        right.fs = fs;
+        left.features = features;
+        right.features = features;
+
+        bool ordinal = features[featureToSplitOn].ordinal;
+
+        double useRatio = 1.0 * rows.size() / rows.back().exampleNum;
+
+        bool reIndex = useRatio < 0.1;
```

Comment:
Nah the idea in time will be to split when the total usable data fits inside cache